src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/Legacy/RoslynCSharpTokenizer.cs

```diff
@@ -462,13 +498,120 @@ private StateResult Trivia()
+        switch (trivia.Kind())
         {
-            CSharpSyntaxKind.WhitespaceTrivia => SyntaxKind.Whitespace,
-            CSharpSyntaxKind.EndOfLineTrivia => SyntaxKind.NewLine,
-            CSharpSyntaxKind.SingleLineCommentTrivia or CSharpSyntaxKind.MultiLineCommentTrivia or CSharpSyntaxKind.MultiLineDocumentationCommentTrivia or CSharpSyntaxKind.SingleLineDocumentationCommentTrivia => SyntaxKind.CSharpComment,
-            var kind => throw new InvalidOperationException($"Unexpected trivia kind: {kind}."),
+            case CSharpSyntaxKind.WhitespaceTrivia:
+                tokenType = SyntaxKind.Whitespace;
+                break;
+            case CSharpSyntaxKind.EndOfLineTrivia:
+                tokenType = SyntaxKind.NewLine;
+                _isOnlyWhitespaceOnLine = true;
+                break;
+            case CSharpSyntaxKind.SingleLineCommentTrivia or
+                 CSharpSyntaxKind.MultiLineCommentTrivia or
+                 CSharpSyntaxKind.MultiLineDocumentationCommentTrivia or
+                 CSharpSyntaxKind.SingleLineDocumentationCommentTrivia:
+                tokenType = SyntaxKind.CSharpComment;
+                _isOnlyWhitespaceOnLine = false;
+                break;
+            case var kind when SyntaxFacts.IsPreprocessorDirective(kind):
+                tokenType = SyntaxKind.CSharpDirective;
+
+                if (!_isOnlyWhitespaceOnLine)
+                {
+                    CurrentErrors.Add(
+                        RazorDiagnosticFactory.CreateParsing_PreprocessorDirectivesMustBeAtTheStartOfLine(
+                            new SourceSpan(CurrentStart, contentLength: trivia.FullSpan.Length)));
+                }
+
+                var directiveTrivia = (DirectiveTriviaSyntax)trivia.GetStructure()!;
+                Debug.Assert(directiveTrivia != null);
+
+                if (directiveTrivia is DefineDirectiveTriviaSyntax or UndefDirectiveTriviaSyntax)
+                {
+                    CurrentErrors.Add(
+                        RazorDiagnosticFactory.CreateParsing_DefineAndUndefNotAllowed(
+                            new SourceSpan(CurrentStart, contentLength: directiveTrivia.FullSpan.Length)));
+                }
+
+                _isOnlyWhitespaceOnLine = directiveTrivia.EndOfDirectiveToken.TrailingTrivia is [.., { RawKind: (int)CSharpSyntaxKind.EndOfLineTrivia }];
+                break;
+            case CSharpSyntaxKind.DisabledTextTrivia:
+                tokenType = SyntaxKind.CSharpDisabledText;
+                _isOnlyWhitespaceOnLine = true;
+
+                // We want to scan through the disabled text and see if someone misplaced an #else or #endif by not putting it at the start of a line. We can't truly
+                // be certain; for example, it could be in html, intentionally. But this is just a warning, and the user can disable it; since we made a breaking change
+                // and there could be directives not at the start of a line, we want to be helpful.
```

Comment:
Consider testing that this warning can be suppressed in source via `#pragma warning disable`.

Reply:
I don't think it can be. I'm not sure any `#pragma warning` directives apply warnings generated directly by the razor compiler currently.